### DISCUSSION TITLE
Extent of GraphQL API?

### DISCUSSION THREAD

**Discussion by bliu-nl** [buildbot/buildbot#7999] [Q&A] [answered]
Hello, we have been happily running Buildbot and querying via the Python API (self.master.data), REST (api/v2/builders), and JavaScript module (DataService & DataAccessor) in our Buildbot and non-Buildbot code. These queries can be quite slow, so we're curious if the GraphQL API can help, but it's hard to find documentation or examples. The main question would be, how much GraphQL support is there and where is the API exposed? Does it have endpoints/clients/etc. that correspond to the queries we already use?

For more context, we've been looking at the GraphQLConnector class at `buildbot/master/buildbot/data/graphql.py` as well as the schema produced by running `buildbot graphql-schema`. GraphQL would be useful for a query like the last 1000 builds, but instead of getting back many attributes per build we're only interested in buildid and started_at. The schema looks like it would support this, although I also don't see an equivalent for REST's properties query (api/v2/builds?property=revision&property=owners). Would we have to do some implementation ourselves to support GraphQL? Our main use case is a custom console view that looks similar to Buildbot's Console View, and does slow queries via the JavaScript DataService. We're looking for ways that GraphQL or other Buildbot features could improve the load times.

**Comment by p12tic** [comment, marked as answer]
graphql is currently experimental. The performance issues of loading e.g. 1000 builds are known, but I didn't have time to prioritize that. In general, the amount of data is so small, that it shouldn't matter how many attributes come together with a build. Most of the time should be spent in the database anyway. If it's not, then it's clear target for optimization.

**Reply by bliu-nl** [reply]
Interesting thank you, in that case we'll focus on trying to reduce the number of builds requested, or the total count of requests. I'll submit a new discussion if I have follow-up questions.

**Reply by p12tic** [reply]
@bliu-nl FYI GraphQL API is slated for removal in near future https://github.com/buildbot/buildbot/pull/8260

Even if the PR does not land in Buildbot 4.2, I think it no longer makes sense to rely on GraphQL being available.

**Reply by bliu-nl** [reply]
Thanks for the update, that's too bad that GraphQL couldn't easily be integrated with the frontend. In the last few months we've started to use the `api/v2/changes/n:changeid/builds` endpoint, and it works well for our use case so far - we can gradually populate a console view with many small queries instead of waiting for one slow builds query.
